frontend/src/components/Table.tsx

```diff
@@ -0,0 +1,113 @@
+import { Pagination, Toolbar, ToolbarContent, ToolbarItem } from '@patternfly/react-core';
+import {
+  TableComposable,
+  Thead,
+  Tr,
+  Th,
+  Tbody,
+  TableComposableProps,
+  Caption,
+  Td,
+} from '@patternfly/react-table';
+import React, { useEffect } from 'react';
+import useTableColumnSort, { SortableData } from '~/utilities/useTableColumnSort';
+
+type TableProps<DataType> = {
+  data: DataType[];
+  columns: SortableData<DataType>[];
+  rowRenderer: (data: DataType, rowIndex: number) => React.ReactNode;
+  enablePagination?: boolean;
+  minPageSize?: number;
+  toolbarContent?: React.ReactElement<typeof ToolbarItem>;
+  emptyTableView?: React.ReactElement<typeof Tr>;
+  caption?: string;
+} & Omit<TableComposableProps, 'ref' | 'data'>;
+
+export const Table = <T,>({
+  data,
+  columns,
+  rowRenderer,
+  enablePagination,
+  minPageSize = 10,
+  toolbarContent,
+  emptyTableView,
+  caption,
+  ...props
+}: TableProps<T>): React.ReactElement => {
+  const [page, setPage] = React.useState(1);
+  const [pageSize, setPageSize] = React.useState(minPageSize);
+
+  // update page to 1 if data changes (common when filter is applied)
+  useEffect(() => setPage(1), [data]);
+
+  const sort = useTableColumnSort<T>(columns, 0);
+
+  const showPagination = enablePagination && data.length > minPageSize;
+  const pagination = (pageDirection: 'up' | 'down') =>
+    showPagination && (
+      <Pagination
+        dropDirection={pageDirection}
+        perPageComponent="button"
+        itemCount={data.length}
+        perPage={pageSize}
+        page={page}
+        onSetPage={(e, newPage) => setPage(newPage)}
+        onPerPageSelect={(e, newSize, newPage) => {
+          setPageSize(newSize);
+          setPage(newPage);
+        }}
+        widgetId="table-pagination"
+      />
+    );
+
+  return (
+    <>
+      <Toolbar>
+        <ToolbarContent>
+          {toolbarContent}
+          <ToolbarItem variant="pagination" alignment={{ default: 'alignRight' }}>
+            {pagination('down')}
+          </ToolbarItem>
+        </ToolbarContent>
+      </Toolbar>
+      <TableComposable {...props}>
+        {caption && <Caption>{caption}</Caption>}
+        <Thead>
+          <Tr>
+            {columns.map((col, i) => (
+              <Th
+                key={col.field + i}
+                sort={col.sortable ? sort.getColumnSort(i) : undefined}
+                width={col.width}
+              >
+                {col.label}
+              </Th>
+            ))}
+          </Tr>
+        </Thead>
+        <Tbody>
+          <>
+            {emptyTableView && data.length === 0 && (
+              <Tr>
+                <Td colSpan={columns.length} style={{ textAlign: 'center' }}>
+                  {emptyTableView}
+                </Td>
+              </Tr>
+            )}
+            {data.map((row, i) => rowRenderer(row, i))}
```

Comment:
You might want to slice the data here based on the pagination settings, or the pagination won't work.
```suggestion
            {data
              .slice(pageSize * (page - 1), pageSize * page)
              .map((row, i) => rowRenderer(row, i))}
```

Reply:
good catch